modules/rgbd/src/colored_kinfu.cpp

```diff
@@ -304,23 +307,27 @@ bool ColoredKinFuImpl<MatType>::updateT(const MatType& _depth, const MatType& _r
         {
             // use depth instead of distance
             volume->integrate(depth, rgb, params.depthFactor, pose, params.intr, params.rgb_intr);
+            newFrame.setPyramidLevel(params.icpIterations.size(), OdometryFramePyramidType::PYR_IMAGE);
+            newFrame.setPyramidAt(rgb, OdometryFramePyramidType::PYR_IMAGE, 0);
         }
         MatType points, normals, colors;
-        newFrame->getPyramidAt(points,  OdometryFrame::PYR_CLOUD, 0);
-        newFrame->getPyramidAt(normals, OdometryFrame::PYR_NORM,  0);
-        newFrame->getPyramidAt(colors,  OdometryFrame::PYR_IMAGE, 0);
+        newFrame.getPyramidAt(points, OdometryFramePyramidType::PYR_CLOUD, 0);
+        newFrame.getPyramidAt(normals, OdometryFramePyramidType::PYR_NORM,  0);
+        newFrame.getPyramidAt(colors, OdometryFramePyramidType::PYR_IMAGE, 0);
+
         volume->raycast(pose, params.intr, params.frameSize, points, normals, colors);
-        //TODO: fix it
-        // This workaround relates to specific process of pyramid building
-        newFrame->setDepth(noArray());
-
-        newFrame->setPyramidAt(points,  OdometryFrame::PYR_CLOUD, 0);
-        newFrame->setPyramidAt(normals, OdometryFrame::PYR_NORM,  0);
-        newFrame->setPyramidAt(colors,  OdometryFrame::PYR_IMAGE, 0);
-        icp->prepareFrameCache(newFrame, OdometryFrame::CACHE_SRC);
+
+        newFrame.setPyramidAt(points, OdometryFramePyramidType::PYR_CLOUD, 0);
+        newFrame.setPyramidAt(normals, OdometryFramePyramidType::PYR_NORM,  0);
+        newFrame.setPyramidAt(colors, OdometryFramePyramidType::PYR_IMAGE, 0);
     }
 
-    prevFrame = newFrame;
+    renderFrame = newFrame;
+
+    prevFrame = icp.createOdometryFrame();
+    newFrame.setImage(rgb);
+    prevFrame.setDepth(depth);
```

Comment:
Why not to do `prevFrame = newFrame`?

Reply:
Oh, looks like colored KinFu code differs from other KinFu pipelines, this should have its reason or be aligned to them.